### PR TITLE
Added alignedAllocate for AscendingPageAllocator

### PR DESCRIPTION
**Pull request by jercaianu**
I also added and refactored some tests for the new functionality.

### CODE REVIEW

File: std/experimental/allocator/building_blocks/ascending_page_allocator.d [hunk 1]
```diff
@@ -101,13 +101,14 @@ public:
     {
         import std.algorithm.comparison : min;
 
+        immutable pagedBytes = numPages * pageSize;
         size_t goodSize = goodAllocSize(n);
-        if (goodSize > numPages * pageSize || offset - data > numPages * pageSize - goodSize)
+        if (goodSize > pagedBytes || offset - data > pagedBytes - goodSize)
```
Comment (andralex) [inline std/experimental/allocator/building_blocks/ascending_page_allocator.d]: `if (goodSize > pagedBytes || offset + goodSize > data + pagedBytes)` would be better from a defensive viewpoint (no subtraction of unsigned values) but you decide whether it's at all sensible :)

File: std/experimental/allocator/building_blocks/ascending_page_allocator.d [hunk 2]
```diff
@@ -101,13 +101,14 @@ public:
...
             return null;
 
         if (offset + goodSize > readWriteLimit)
```
Comment (andralex) [inline std/experimental/allocator/building_blocks/ascending_page_allocator.d]: ah, so here we have again `offset + goodSize` as a meaningful quantity

File: std/experimental/allocator/building_blocks/ascending_page_allocator.d [hunk 3]
```diff
@@ -144,6 +145,38 @@ public:
...
+    `null` on failure or if the requested size exceeds the remaining capacity.
+    */
+    void[] alignedAllocate(size_t n, uint a)
+    {
+        void* alignedStart = cast(void*) roundUpToMultipleOf(cast(size_t) offset, a);
+        assert(alignedStart.alignedAt(a));
+        immutable pagedBytes = numPages * pageSize;
+        size_t goodSize = goodAllocSize(n);
+        if (goodSize > pagedBytes ||
+            alignedStart - data > pagedBytes - goodSize)
```
Comment (andralex) [inline std/experimental/allocator/building_blocks/ascending_page_allocator.d]: one more indent here or full block

File: std/experimental/allocator/building_blocks/ascending_page_allocator.d (hunk 1 above)
Comment (jercaianu) [reply std/experimental/allocator/building_blocks/ascending_page_allocator.d]: When I did this, I was afraid that some sort of an overflow could happen.
By doing substraction and knowing that ``offset > data`` and ``pagedBytes > goodSize`` I didnt have to do any overflow checks.

**Pull request merged**